autoendpoint/src/routers/apns/router.rs

```diff
@@ -0,0 +1,536 @@
+                self.incr_error_metric("unknown", channel);
+            }
+        }
+
+        ApiError::from(ApnsError::ApnsUpstream(error))
+    }
+}
+
+#[async_trait(?Send)]
+impl Router for ApnsRouter {
+    fn register(
+        &self,
+        router_input: &RouterDataInput,
+        app_id: &str,
+    ) -> Result<HashMap<String, Value>, RouterError> {
+        if !self.clients.contains_key(app_id) {
+            return Err(ApnsError::InvalidReleaseChannel.into());
+        }
+
+        let mut router_data = HashMap::new();
+        router_data.insert(
+            "token".to_string(),
+            serde_json::to_value(&router_input.token).unwrap(),
+        );
+        router_data.insert(
+            "rel_channel".to_string(),
+            serde_json::to_value(app_id).unwrap(),
+        );
+
+        if let Some(aps) = &router_input.aps {
+            if APS::deserialize(aps).is_err() {
+                return Err(ApnsError::InvalidApsData.into());
+            }
+
+            router_data.insert("aps".to_string(), aps.clone());
+        }
+
+        Ok(router_data)
+    }
+
+    async fn route_notification(&self, notification: &Notification) -> ApiResult<RouterResponse> {
+        debug!(
+            "Sending APNS notification to UAID {}",
+            notification.subscription.user.uaid
+        );
+        trace!("Notification = {:?}", notification);
+
+        // Build message data
+        let router_data = notification
+            .subscription
+            .user
+            .router_data
+            .as_ref()
+            .ok_or(ApnsError::NoDeviceToken)?;
+        let token = router_data
+            .get("token")
+            .and_then(Value::as_str)
+            .ok_or(ApnsError::NoDeviceToken)?;
+        let channel = router_data
+            .get("rel_channel")
+            .and_then(Value::as_str)
+            .ok_or(ApnsError::NoReleaseChannel)?;
+        let aps: APS<'_> = router_data
+            .get("aps")
+            .and_then(|value| APS::deserialize(value).ok())
+            .unwrap_or_else(Self::default_aps);
+        let mut message_data = build_message_data(notification, self.settings.max_data)?;
+        message_data.insert("ver", notification.message_id.clone());
+
+        // Get client and build payload
+        let ApnsClientData { client, topic } = self
+            .clients
+            .get(channel)
+            .ok_or(ApnsError::InvalidReleaseChannel)?;
+        let payload = Payload {
+            aps,
+            data: message_data
+                .into_iter()
+                .map(|(k, v)| (k, Value::String(v)))
+                .collect(),
+            device_token: token,
+            options: NotificationOptions {
+                apns_id: None,
+                apns_priority: Some(Priority::High),
+                apns_topic: Some(topic),
+                apns_collapse_id: None,
+                apns_expiration: None,
```

Comment:
hrm, this should probably be the message TTL. We don't want messages appearing after they've expired. (I'll file a separate issue for that since the python never got updated either.)

Reply:
https://github.com/mozilla-services/autopush-rs/issues/202

Reply:
This is easily fixed in this PR. Can we move the issue to the Python repo instead?

Reply:
Sure. done.